distributed/tests/test_worker_state_machine.py

```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
+            stimulus_id="failure",
+        ),
+    )
+    assert ws.tasks["x"].state == "error"
+    assert ws.available_resources == {"R": 1}
+
+
+@pytest.mark.parametrize("state", ["executing", "long-running"])
+def test_resource_restricted_task_with_dependencies(ws, state):
+    ws.available_resources = {"R": 1}
+    ws.total_resources = {"R": 1}
+
+    instructions = ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            key="y",
+            who_has={"x": ["127.0.0.1:1235"]},
+            nbytes={"x": 8},
+            resource_restrictions={"R": 1},
+            stimulus_id="compute",
+        )
+    )
+    assert ws.tasks["x"].state == "flight"
+    assert ws.tasks["y"].state == "waiting"
+    assert ws.available_resources == {"R": 1}
+
+    ws.handle_stimulus(
+        GatherDepSuccessEvent("gather-dep-done", "127.0.0.1:1235", 8, {"x": 1.0})
+    )
+    # Ensure that y transitions through the ``constrained`` state
+    expected_story = [
+        ("y", "compute-task", "released"),
+        ("y", "released", "waiting", "waiting", {"x": "fetch"}),
+        ("y", "waiting", "ready", "waiting", {"y": "constrained"}),
+        ("y", "waiting", "constrained", "constrained", {"y": "executing"}),
+        ("y", "constrained", "executing", "executing", {}),
+    ]
+
+    if state == "long-running":
+        ws.handle_stimulus(
+            SecedeEvent(key="y", compute_duration=1.0, stimulus_id="secede")
+        )
+        expected_story.append(("y", "executing", "long-running", "long-running", {}))
+
+    assert ws.tasks["x"].state == "memory"
+    assert ws.tasks["y"].state == state
+    assert ws.available_resources == {"R": 0}
+    assert_story(
+        ws.story("y"),
+        expected_story,
+        strict=True,
+    )
+
+    ws.handle_stimulus(
+        ExecuteSuccessEvent(
+            key="y",
+            value=None,
+            start=0.0,
+            stop=1.0,
+            nbytes=8,
+            type=None,
+            stimulus_id="success",
+        ),
+    )
+    assert ws.tasks["y"].state == "memory"
+    assert ws.available_resources == {"R": 1}
+
+
+def test_resumed_executing_task_releases_resources_when_done(ws_with_running_task):
+    ws = ws_with_running_task
+
+    ws.handle_stimulus(FreeKeysEvent("cancel", ["x"]))
+    assert ws.tasks["x"].state == "cancelled"
+    assert ws.available_resources == {"R": 0}
+
+    ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            key="y",
+            who_has={"x": ["127.0.0.1:1235"]},
+            nbytes={"x": 8},
+            stimulus_id="compute",
+        )
+    )
+    ws.handle_stimulus(
+        ExecuteSuccessEvent(
+            key="x",
+            value=None,
+            start=0.0,
+            stop=1.0,
+            nbytes=8,
+            type=None,
+            stimulus_id="s2",
+        ),
+    )
+    assert ws.tasks["x"].state == "memory"
+    assert ws.available_resources == {"R": 1}
+
+
+@pytest.mark.xfail(reason="distributed#6682")
```

Comment:
`XFAIL` to be tackled in a follow-up PR. #6682 has been created to tackle this issue.